### PR TITLE
RPC & Mempool back ports.

### PR DESCRIPTION
**Pull request by furszy**
More back ports and adaptations over the RPC values parsing and mempool. We need to get closer in this area :) .

* bitcoin#6379
* bitcoin#6715
* bitcoin#6915
* bitcoin#7007
* bitcoin#7008
* bitcoin#12643
* bitcoin#18474
* bitcoin#18704

### CODE REVIEW

**Comment by random-zebra** [comment]
Would be good to have bitcoin@fa1436c42958688bcc48418f5b434a80a53fd5ac and bitcoin@fac2fc4dd8a28b99e17c57e4ab6580a3231f1d0a fully backported here (i.e. not only `sync_blocks()`, but also `sync_mempools()`)

**Comment by furszy** [comment]
@random-zebra done.
We still need to back port `syncwithvalidationinterfacequeue` RPC command in a future PR.

**Comment by furszy** [comment]
conflicts solved. Rebased on top of master.

**Comment by furszy** [comment]
Solved all the comments. Ready for another round of review.

File: src/miner.cpp [hunk 1]
```diff
@@ -568,7 +569,12 @@ void BitcoinMiner(CWallet* pwallet, bool fProofOfStake)
     const int64_t nSpacingMillis = consensus.nTargetSpacing * 1000;
 
     // Each thread has its own key and counter
-    CReserveKey reservekey(pwallet);
+    Optional<CReserveKey> opReservekey{nullopt};
```
Comment (Fuzzbawls) [inline src/miner.cpp:572]: Nit: should probably include our `optional.h` wrapper directly in the header includes rather than implicitly

**Pull request merged**